File: ProcessoConcrete/app/src/main/java/br/com/guilherme/concrete/presenter/PullRequestPresenter.java
package br.com.guilherme.concrete.presenter;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import br.com.guilherme.concrete.model.PullRequest;
import br.com.guilherme.concrete.model.RepositorioService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PullRequestPresenter {
    private View view;
    private RepositorioService service;

    public PullRequestPresenter(View view) {
        this.view = view;
        service = new RepositorioService();
    }

    public void getAllPulls() {
        service.getAllPulls()
                .enqueue(new Callback<List<PullRequest>>() {
                    @Override
                    public void onResponse(Call<List<PullRequest>> call, Response<List<PullRequest>> response) {
                        List<PullRequest> pullRequests = new ArrayList<>();
                        if (response.body() != null)
                            pullRequests.addAll(response.body());
                        view.setRecyclerView(pullRequests);
                    }

                    @Override
                    public void onFailure(Call<List<PullRequest>> call, Throwable t) {
                        view.onError(t.getMessage());
                        Log.e("api_erro", t.getMessage());
                    }
                });
    }

    public interface View{
        void setRecyclerView(List<PullRequest> repositorios);
        void onError(String errorException);
    }
}

File: app/src/main/java/wilquer/lima/desafioconcrete/ui/pullrequest/PullRequestPresenter.kt
package wilquer.lima.desafioconcrete.ui.pullrequest

import org.jetbrains.anko.doAsync
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import wilquer.lima.desafioconcrete.data.model.PullRequest
import wilquer.lima.desafioconcrete.network.RetrofitApi
import wilquer.lima.desafioconcrete.network.service.PullRequestService
import wilquer.lima.desafioconcrete.util.Constants

class PullRequestPresenter(val view: PullRequestContract.View) : PullRequestContract.Presenter {

    override fun getPullRequests(creator: String, repo: String) {
        view.setProgress(true)

        doAsync {
            val apiService = RetrofitApi(Constants.GENERAL_URL).client.create(PullRequestService::class.java)

            val call = apiService.getPullRequests(creator, repo)
            call.enqueue(object : Callback<List<PullRequest>>{
                override fun onFailure(call: Call<List<PullRequest>>, t: Throwable) {
                    view.setProgress(false)
                    view.error(t.message.toString())
                }

                override fun onResponse(call: Call<List<PullRequest>>, response: Response<List<PullRequest>>) {
                    view.setProgress(false)
                    view.pullrequests(response.body())
                }

            })
        }
    }


}